clients/src/main/java/org/apache/kafka/common/serialization/ListDeserializer.java

```diff
@@ -0,0 +1,105 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+import com.sun.xml.internal.ws.encoding.soap.DeserializationException;
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ListDeserializer<T> implements Deserializer<List<T>> {
+
+    private Deserializer<T> inner;
+    private Class listClass;
+    private Integer primitiveSize;
+
+    private Map<Class, Integer> primitiveDeserializers = new HashMap<Class, Integer>() {{
+        put(LongDeserializer.class, 8);
+        put(IntegerDeserializer.class, 4);
+        put(ShortDeserializer.class, 2);
+        put(FloatDeserializer.class, 4);
+        put(DoubleDeserializer.class, 8);
+    }};
+
+    public ListDeserializer() {
+    }
+
+    public ListDeserializer(Class listClass, Deserializer<T> deserializer) {
+        this.listClass = listClass;
+        this.inner = deserializer;
+        this.primitiveSize = primitiveDeserializers.get(deserializer.getClass());
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (inner == null) {
+            String listTypePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_TYPE_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_TYPE_CLASS;
+            String innerSerdePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_INNER_CLASS;
+            String listType = (String) configs.get(listTypePropertyName);
+            String innerSerde = (String) configs.get(innerSerdePropertyName);
+            try {
+                listClass = Class.forName(listType);
+                this.inner = Utils.newInstance(innerSerde, Serde.class).deserializer();
+                inner.configure(configs, isKey);
+            } catch (ClassNotFoundException e) {
+                throw new DeserializationException("Could not find a class for \"" + listType + "\"", e);
+            }
+        }
+    }
+
+    private List<T> getListInstance(int listSize) {
+        try {
+            Constructor<?> listConstructor = listClass.getConstructor(Integer.TYPE);
```

Comment:
How do we know that all list types implement a constructor like this? Should we have a fall back to default constructor?

Reply:
I guess smth like this:
```
try {
            Constructor<?> listConstructor;
            try {
                listConstructor = listClass.getConstructor(Integer.TYPE);
                return (List<T>) listConstructor.newInstance(listSize);
            } catch (NoSuchMethodException e) {
                listConstructor = listClass.getConstructor();
                return (List<T>) listConstructor.newInstance();
            }
        } catch (Exception e) {
            throw new RuntimeException("Could not construct a list instance of \"" + listClass.getCanonicalName() + "\"", e);
        }
```

Reply:
Probably add a warning log? What do you think?

Reply:
Don't think we need a warning. Wondering if we should use try-catch or better pro-actively check if an int-constructor exists?

Reply:
We kind of implicitly check if int-constructor exists using this try-catch block, right?